### PR TITLE
feat: add pseudo single file support

### PR DESCRIPTION
**Pull request by mjlbach**
@lithammer I am finally doing it. Victory shall be mine.

### CODE REVIEW

**Comment by lithammer** [comment]
> @lithammer I am finally doing it. Victory shall be mine.

Don't know if I should be scared or happy 😜

File: lua/lspconfig/jedi_language_server.lua [hunk 1]
```diff
@@ -1,12 +1,21 @@
 local configs = require 'lspconfig/configs'
+local util = require 'lspconfig/util'
+
+local root_files = {
+  'pyproject.toml',
+  'setup.py',
+  'setup.cfg',
+  'requirements.txt',
+  'Pipfile',
+  'pyrightconfig.json',
```
Comment (lithammer) [inline lua/lspconfig/jedi_language_server.lua:10]:
```suggestion
```

**Comment by mjlbach** [comment]
Happy because this should "just work" at least with the servers I tested... I thought about adding alternative modes (one that falls back to dirname for servers that require a root directory in order to launch like metals, one that attaches to a common scratch server per all languages with an nil root/root in tmp, etc.) but I'd rather push upstream to add something official to the spec.

**Comment by clason** [comment]
Could you explain what this change _does_? How do servers behave now?

**Comment by mjlbach** [comment]
Sure, I haven't documented it yet because I want testers.

If a server fails to find a root directory, if `single_file_mode` is enabled, it will start the server anyways, just without sending the root. This works fine on the subset of servers I enabled it on, but crashes some servers (like metals), so I made it an opt-in flag. These servers are tracked in a separate table than servers which share the root, so single file mode servers can share a server if it's in the same directory (all lua files in home will share a server in home), but not across directories.

**Comment by lithammer** [comment]
> [...] but I'd rather push upstream to add something official to the spec.

I think that's wise. The fewer workarounds we have to do in the client(s), the better.

**Review by lithammer — APPROVED** [review]

**Comment by mjlbach** [comment]
I'm going to hold off on merging this until others have tested (@clason was concerned this might break texlab in multi-file projects).

**Comment by clason** [comment]
Don't hold off on my account. (I still think for texlab specifically, it's simpler just to default to `null` -- i.e., _not_ setting a `root_directory` in the default config, but that shouldn't block this PR.)

**Comment by mjlbach** [comment]
I'm planning on merging this PR tonight unless anyone provides me a concrete failure case. Ofc root directories can be updated in the future.

File: lua/lspconfig/configs.lua [hunk 2]
```diff
@@ -78,22 +78,29 @@ function configs.__newindex(t, config_name, config_def)
...
+            M.manager.try_add_wrapper(bufnr)
+          end
         end
+      elseif not root_dir and config.single_file_support then
+        -- This allows on_new_config to use the parent directory of the file
+        -- Effectively this is the root from lspconfig's perspective, as we use
+        -- this to attach additional files in the same parent folder to the same server.
+        -- We just no longer send rootDirectory or workspaceFolders during initialization.
+        local pseudo_root = util.path.dirname(api.nvim_buf_get_name(0))
+        M.manager.add(pseudo_root, true)
```
Comment (kylo252) [inline lua/lspconfig/configs.lua:101]: Can you not have multiple buffers for the same file, with splits or something? I feel like that isn't covered here.

**Review by kylo252 — CHANGES_REQUESTED** [review]

File: lua/lspconfig/configs.lua (hunk 2 above)
Comment (mjlbach) [reply lua/lspconfig/configs.lua:101]: did you try it :)

File: lua/lspconfig/util.lua [hunk 3]
```diff
@@ -265,9 +271,19 @@ function M.server_per_root_dir_manager(_make_config)
       end
       new_config.on_exit = M.add_hook_before(new_config.on_exit, function()
         clients[root_dir] = nil
+        single_file_clients[root_dir] = nil
       end)
+      -- TODO This is a hack, this isn't codified in the spec and breaks many servers.
```
[the remark this answers is not in the document]
Comment (mjlbach) [reply lua/lspconfig/util.lua:276]: I use TODO scanning for my own tracking

File: lua/lspconfig/configs.lua (hunk 2 above)
Comment (kylo252) [reply lua/lspconfig/configs.lua:101]: I did and it seems to work fine 🚀

Comment (mjlbach) [reply lua/lspconfig/configs.lua:101]: I know ;)

File: lua/lspconfig/util.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (kylo252) [reply lua/lspconfig/util.lua:276]: I just thought it might be easier to spot, it's sort of standardized.

[the remark this answers is not in the document]
Comment (clason) [reply lua/lspconfig/util.lua:276]: [citation needed]

[the remark this answers is not in the document]
Comment (mjlbach) [reply lua/lspconfig/util.lua:276]: [1] "This breaks at least scala of the servers I tested"
                                                                     - Michael Lingelbach, Chief LSP officer of Neovim

[the remark this answers is not in the document]
Comment (mjlbach) [reply lua/lspconfig/util.lua:276]: > The workspace/workspaceFolders request is sent from the server to the client to fetch the current open list of workspace folders. Returns null in the response if only a single file is open in the tool. Returns an empty array if a workspace is open but no folders are configured.

Workspace folders, added in 3.16, implies that in response to a request the workspace folders can be null if a single file is open, however for rootURI it's at best unclear (deprecated, but still used by some servers), it's not explicitly codified in the initializationParams section https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#initializeParams.

I'm fine if in 3.17 they say "send nil root on initialization if a single file is open", bc that makes my life easier.

[the remark this answers is not in the document]
Comment (kylo252) [reply lua/lspconfig/util.lua:276]: How about setting this to `false`?
```lua
 workspace = {
        applyEdit = true,
        configuration = true,
        symbol = {
          dynamicRegistration = false,
          hierarchicalWorkspaceSymbolSupport = true,
          symbolKind = {
            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }
          }
        },
        workspaceEdit = {
          resourceOperations = { "rename", "create", "delete" }
        },
        workspaceFolders = true --> this one
      }
```

[the remark this answers is not in the document]
Comment (mjlbach) [reply lua/lspconfig/util.lua:276]: ... no

**Comment by aktau** [comment]
So this broke one of my [workflows](https://xkcd.com/1172/) (I think): https://github.com/aktau/dotfiles/blob/1e44e7760dd000fa86612fc5805e5e15fc1905f5/.vim/lua/lsp.lua#L277.

Basically, I want to **inhibit** the start of certain LSP servers for certain paths. Said another way, I want (for the same language) the following to happen:

```
/my/special/root/path.c -> mycustomlsp
/* -> clangd
```

Given that clangd apparently recently gained support for single-file mode, this no longer works. I'm sure I've been abusing this "feature" of nvim-lspconfig, but I can't figure out how to do it the right way. Any tips?

**Comment by mjlbach** [comment]
Sure, just override single file support in setup and you go back to the old behavior

**Comment by aktau** [comment]
> Sure, just override single file support in setup and you go back to the old behavior. 

I don't necessarily want to get rid of single file support, I guess it could be useful. I just wanted to prevent certain LSPs from running if the project is in a certain directory.

I checked how `vim.lsp.start_client` was called from `nvim-lspconfig` and realized that I could use the `on_new_config` hook to get a similar effect as what I used to have by setting `config.enabled` to `false`.

```patch
diff --git a/.vim/lua/lsp.lua b/.vim/lua/lsp.lua                                                                                                                                                                                                                                      
index 83f2d1f..65de859 100644                                                                                                                                                                                                                                                         
--- a/.vim/lua/lsp.lua                                                                                                                                                                                                                                                                
+++ b/.vim/lua/lsp.lua                                                                                                                                                                                                                                                                
@@ -271,13 +271,12 @@ for lsp, config in pairs(servers) do                                                                                                                                                                                                                            
    config,                                                                                                                                                                                                                                                                          
     -- Ensure the regular LSPs do not run in the overridden roots.                                                                                                                                                                                                                   
     {                                                                                                                                                                                                                                                                                
-      root_dir = function(fname)                                                                                                                                                                                                                                                     
+      on_new_config = function(config, root_dir)                                                                                                                                                                                                                                     
         -- TODO: don't install the override if the filetypes don't match (so                                                                                                                                                                                                         
         --       that efm-langserver can work in all scenarios).                                                                                                                                                                                                                     
-        if override_lsp_root(fname) ~= nil then                                                                                                                                                                                                                                      
-          return nil                                                                                                                                                                                                                                                                 
+        if override_lsp_root(root_dir) ~= nil then                                                                                                                                                                                                                                   
+          config.enabled = false                                                                                                                                                                                                                                                     
        end                                                                                                                                                                                                                                                                          
-        return root_dir_fn(fname)                                                                                                                                                                                                                                                    
       end,                                                                                                                                                                                                                                                                           
     }                                                                                                                                                                                                                                                                                
   ))   
```

**Comment by mjlbach** [comment]
Also to be clear (even though you have a workaround) single file support can be selectively disabled per server

**Comment by aktau** [comment]
Yes, I did consider overriding this for every LSP to false. But as I said, it may end up being useful outside of the overridden folders when I'm hacking on something.

Side note: for `efm-langserver` I've always worked around the problem that I think single-file support mode is trying to solve by just returning `'/'` as the root. Never had any problems with it.

**Comment by mjlbach** [comment]
> Yes, I did consider overriding this for every LSP to false. 

To clarify I was suggesting to do this only for servers that you wanted to use the root trick on.

> I've always worked around the problem that I think single-file support mode is trying to solve by just returning '/' as the root. 

Sending inaccurate (non-project specific) root directories like '/' *is* the problem single file support mode is trying to solve. Some servers (pyright/bashls) index everything under root. If we can't find a root, instead of starting or not starting the server, we start the server without the root marker (in many cases preventing expensive indexing) and attach the buffer

**Comment by dstein64** [comment]
I call `setup` with `autostart = false`, and then use `LspStart` to start a language server.

Prior to this PR, the LSP client would automatically attach to the active buffer.

After this PR, I have to manually call `vim.lsp.buf_attach_client` after the call to `LspStart`.

Is this intentional and if so is there a setting I can use to retain the existing behavior? Thanks!

**Comment by kylo252** [comment]
> Is the behavior—not attaching to the active buffer after `LspStart` when `autostart = false` and LSP is running in single-file mode—intentional? 

I don't think it's intentional, since `LspStart` should launch and _attach_ a server. If nothing else, we can make it clear that it's not supported.

So you should probably open a new ticket.

**Pull request merged**